File: swift_module/tunnel/OutlineInteractor.swift
import NetworkExtension
import CommonDI
import MyLibrary
import os
import app
import Foundation
import SystemConfiguration
import Network

public final class OutlineInteractor {
    private var logs = NativeModuleHolder.logsRepository
    private var outlineStarted: Bool = false

    func startOutline(
        tunnelFileDescriptor: Int32,
        mtu: Int,
        nativeClientCreated: () -> Void
    ) throws {
        let start = Date()
        logs.writeLog(log: "[Outline] startOutline begin fd=\(tunnelFileDescriptor) mtu=\(mtu)")

        let methodPassword = configsRepository.getMethodPasswordOutline()
        let serverPort = configsRepository.getServerPort()
        let prefix = configsRepository.getPrefixOutline()
        let websocketEnabled = configsRepository.getIsWebsocketEnabled()
        let tcpPath = configsRepository.getTcpPathOutline()
        let udpPath = configsRepository.getUdpPathOutline()
        logs.writeLog(
            log: "Config snapshot: serverPort.len=\(serverPort.count) methodPassword.len=\(methodPassword.count) " +
            "ws=\(websocketEnabled) tcpPath.len=\(tcpPath.count) udpPath.len=\(udpPath.count) " +
            "tunnelFD=\(tunnelFileDescriptor) mtu=\(mtu)"
        )

        // Validate config early (prevents passing empty config into native layer).
        if methodPassword.isEmpty || serverPort.isEmpty {
            logs.writeLog(log: "[startTunnel] Empty Outline config: methodPassword.isEmpty=\(methodPassword.isEmpty) serverPort.isEmpty=\(serverPort.isEmpty) → abort")
            throw NSError(
                domain: "PacketTunnelProvider",
                code: -2,
                userInfo: [NSLocalizedDescriptionKey: "Empty Outline configuration"]
            )
        }

        let config = buildOutlineConfig(
            methodPassword: methodPassword,
            serverPort: serverPort,
            prefix: prefix,
            websocketEnabled: websocketEnabled,
            tcpPath: tcpPath,
            udpPath: udpPath
        )
        logs.writeLog(log: "Outline config built (prefix=\(!prefix.isEmpty), ws=\(websocketEnabled), tcpPath=\(!tcpPath.isEmpty), udpPath=\(!udpPath.isEmpty))")
        if websocketEnabled {
            logs.writeLog(
                log: "WebSocket transport requested (wss) " +
                    "serverHost=\(maskStr(value: OutlineInteractor.extractHost(from: serverPort)))"
            )
        }

        var err: NSError?

        logs.writeLog(log: "[DEBUG][Outline] calling native NewVpnClient protocol=outline fd=\(tunnelFileDescriptor) mtu=\(mtu)")
        let newClientStart = Date()
        Cloak_outlineNewVpnClient(config, "outline", Int(tunnelFileDescriptor), mtu, &err)
        if let error = err {
            logs.writeLog(
                log: "[Outline] NewVpnClient failed in \(elapsedMs(since: newClientStart))ms: " +
                    "\(error.localizedDescription)"
            )
            throw error
        }
        logs.writeLog(log: "[Outline] NewVpnClient succeeded in \(elapsedMs(since: newClientStart))ms")
        nativeClientCreated()
        logs.writeLog(log: "[Outline] nativeClientCreated callback returned")

        logs.writeLog(log: "[DEBUG][Outline] calling native VpnConnect")
        let connectStart = Date()
        Cloak_outlineVpnConnect(&err)
        if let error = err {
            logs.writeLog(
                log: "[Outline] VpnConnect failed in \(elapsedMs(since: connectStart))ms: " +
                    "\(error.localizedDescription)"
            )
            throw error
        }
        outlineStarted = true
        logs.writeLog(
            log: "[Outline] VpnConnect succeeded in \(elapsedMs(since: connectStart))ms " +
                "totalStartMs=\(elapsedMs(since: start))"
        )
    }

    func stopOutline() throws {
        var err: NSError?

        logs.writeLog(log: "[DEBUG][Outline] calling native VpnDisconnect outlineStarted=\(outlineStarted)")
        let start = Date()
        Cloak_outlineVpnDisconnect(&err)
        if let error = err {
            logs.writeLog(
                log: "[Outline] VpnDisconnect failed in \(elapsedMs(since: start))ms: " +
                    "\(error.localizedDescription)"
            )
            throw error
        }
        outlineStarted = false
        logs.writeLog(log: "[DEBUG][Outline] VpnDisconnect returned in \(elapsedMs(since: start))ms")
    }

    func outlineStatus() -> String {
        var err: NSError?
        let start = Date()
        logs.writeLog(log: "[Outline] VpnStatus begin outlineStarted=\(outlineStarted)")
        let status = Cloak_outlineVpnStatus(&err)
        if let error = err {
            let message = "client=true localProxyAlive=false statusError=\(error.localizedDescription)"
            logs.writeLog(
                log: "[Outline] VpnStatus failed in \(elapsedMs(since: start))ms: " +
                    "\(error.localizedDescription)"
            )
            return message
        }
        logs.writeLog(log: "[Outline] VpnStatus returned in \(elapsedMs(since: start))ms status=\(status)")
        return status
    }

    func buildOutlineConfig(
        methodPassword: String,
        serverPort: String,
        prefix: String = "",
        websocketEnabled: Bool = false,
        tcpPath: String = "",
        udpPath: String = ""
    ) -> String {
        let encoded = methodPassword.data(using: .utf8)?.base64EncodedString() ?? ""
        let baseUrl = "ss://\(encoded)@\(serverPort)"

        func extractHost(_ hostPortMaybeWithQuery: String) -> String {
            let hostPort = hostPortMaybeWithQuery.split(separator: "?", maxSplits: 1, omittingEmptySubsequences: true).first.map(String.init) ?? hostPortMaybeWithQuery
            let trimmed = hostPort.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.hasPrefix("[") {
                // IPv6 wrapped in square brackets: [2001:db8::1]:443
                if let start = trimmed.firstIndex(of: "["), let end = trimmed.firstIndex(of: "]"), start < end {
                    return String(trimmed[trimmed.index(after: start)..<end])
                }
            }
            if let lastColon = trimmed.lastIndex(of: ":"), trimmed.filter({ $0 == ":" }).count == 1 {
                return String(trimmed[..<lastColon])
            }
            return trimmed
        }

        // Add the `prefix` query param if present (URL-encoded)
        let ssUrl: String
        if !prefix.isEmpty {
            let separator = serverPort.contains("?") ? "&" : "?"
            let encodedPrefix = prefix.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? prefix
            ssUrl = "\(baseUrl)\(separator)prefix=\(encodedPrefix)"
        } else {
            ssUrl = baseUrl
        }

        // If WebSocket is enabled, wrap the Shadowsocks URL into WebSocket-over-TLS transport (wss://)
        if websocketEnabled {
            let effectiveHost = extractHost(serverPort).trimmingCharacters(in: .whitespacesAndNewlines)
            logs.writeLog(
                log: "[Outline] building WSS config effectiveHost=\(maskStr(value: effectiveHost)) " +
                    "tcpPath.len=\(tcpPath.count) udpPath.len=\(udpPath.count)"
            )

            var wsParams: [String] = []
            // outline-sdk v0.0.16: ws: does not support the `host=` option ("Unsupported option host")
            // Domain-related routing should be configured via TLS SNI (tls:sni=...).
            if !tcpPath.isEmpty { wsParams.append("tcp_path=\(tcpPath)") }
            if !udpPath.isEmpty { wsParams.append("udp_path=\(udpPath)") }

            let wsParamsStr = wsParams.joined(separator: "&")
            // Use tls:sni|ws: for WebSocket-over-TLS (wss://) via SNI
            let tlsPrefix = "tls:sni=\(effectiveHost)"
            if !wsParamsStr.isEmpty {
                return "\(tlsPrefix)|ws:\(wsParamsStr)|\(ssUrl)"
            } else {
                return "\(tlsPrefix)|ws:|\(ssUrl)"
            }
        } else {
            return ssUrl
        }
    }

    public static func extractHost(from hostPortMaybeWithQuery: String) -> String {
        let hostPort = hostPortMaybeWithQuery.split(separator: "?", maxSplits: 1, omittingEmptySubsequences: true).first.map(String.init) ?? hostPortMaybeWithQuery
        let trimmed = hostPort.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix("[") {
            // IPv6 in brackets: [2001:db8::1]:443
            if let start = trimmed.firstIndex(of: "["), let end = trimmed.firstIndex(of: "]"), start < end {
                return String(trimmed[trimmed.index(after: start)..<end])
            }
        }
        if let lastColon = trimmed.lastIndex(of: ":"), trimmed.filter({ $0 == ":" }).count == 1 {
            return String(trimmed[..<lastColon])
        }
        return trimmed
    }

    private func elapsedMs(since start: Date) -> Int {
        Int(Date().timeIntervalSince(start) * 1000)
    }
}

File: kmp_module/app/src/androidMain/kotlin/com/dobby/feature/vpn_service/domain/outline/OutlineInteractor.kt
package com.dobby.feature.vpn_service.domain.outline

import com.dobby.feature.logging.Logger
import com.dobby.feature.main.domain.DobbyConfigsRepository
import com.dobby.feature.vpn_service.DobbyVpnService
import com.dobby.feature.vpn_service.OutlineLibFacade
import com.dobby.feature.vpn_service.domain.descriptor.FDManager
import java.util.*

private fun extractHostFromHostPort(hostPortMaybeWithQuery: String): String {
    val hostPort = hostPortMaybeWithQuery.substringBefore("?").trim()
    if (hostPort.startsWith("[")) {
        // IPv6 in brackets: [2001:db8::1]:443
        return hostPort.substringAfter("[").substringBefore("]")
    }
    // host:port (best-effort)
    val lastColon = hostPort.lastIndexOf(':')
    return if (lastColon > 0 && hostPort.count { it == ':' } == 1) {
        hostPort.substring(0, lastColon)
    } else {
        hostPort
    }
}

private fun buildOutlineUrl(
        methodPassword: String,
        serverPort: String,
        prefix: String = "",
        websocketEnabled: Boolean = false,
        tcpPath: String = "",
        udpPath: String = ""
): String {
    val encoded = Base64.getEncoder().encodeToString(methodPassword.toByteArray())
    val baseUrl = "ss://$encoded@$serverPort"

    // Add prefix parameter if present (URL-encoded)
    val ssUrl = if (prefix.isNotEmpty()) {
        val separator = if (serverPort.contains("?")) "&" else "?"
        val encodedPrefix = java.net.URLEncoder.encode(prefix, "UTF-8")
        "$baseUrl${separator}prefix=$encodedPrefix"
    } else {
        baseUrl
    }

    // Wrap with WebSocket over TLS transport if enabled (wss://)
    val result = if (websocketEnabled) {
        val effectiveHost = extractHostFromHostPort(serverPort).trim()
        val wsParams = buildList {
            if (tcpPath.isNotEmpty()) add("tcp_path=$tcpPath")
            if (udpPath.isNotEmpty()) add("udp_path=$udpPath")
        }.joinToString("&")

        // Use tls:sni|ws: for WebSocket over TLS (wss://) with SNI
        val tlsPrefix = "tls:sni=$effectiveHost"
        if (wsParams.isNotEmpty()) {
            "$tlsPrefix|ws:$wsParams|$ssUrl"
        } else {
            "$tlsPrefix|ws:|$ssUrl"
        }
    } else {
        ssUrl
    }

    return result
}

class OutlineInteractor(
    private val logger: Logger,
    private val dobbyConfigsRepository: DobbyConfigsRepository,
    private val outlineLibFacade: OutlineLibFacade,
    private val fdManager: FDManager,
) {
    private val interfaceFactory = OutlineVpnInterfaceFactory(logger)

    fun startOutline(dobbyVpnService: DobbyVpnService?): Boolean {
        val serviceId = dobbyVpnService?.serviceId ?: "unknown"
        logger.log("[svc:$serviceId] startCloakOutline(): lock acquired vpnInterface=${dobbyVpnService?.vpnInterface?.fd}")
        val shouldTurnOutlineOn = dobbyConfigsRepository.getIsOutlineEnabled()
        logger.log("[svc:$serviceId] startCloakOutline(): shouldTurnOutlineOn=$shouldTurnOutlineOn")

        if (!shouldTurnOutlineOn) {
            logger.log("Start disconnecting Outline")
            return false
        }

        val methodPassword = dobbyConfigsRepository.getMethodPasswordOutline()
        val serverPort = dobbyConfigsRepository.getServerPort()
        val prefix = dobbyConfigsRepository.getPrefixOutline()
        val websocketEnabled = dobbyConfigsRepository.getIsWebsocketEnabled()
        val tcpPath = dobbyConfigsRepository.getTcpPathOutline()
        val udpPath = dobbyConfigsRepository.getUdpPathOutline()
        logger.log("DEBUG: tcpPath='$tcpPath', udpPath='$udpPath'")

        if (methodPassword.isEmpty() || serverPort.isEmpty()) {
            logger.log("Previously used outline apiKey is empty")
            return false
        }

        logger.log("Start connecting Outline")
        val outlineUrl = buildOutlineUrl(
            methodPassword = methodPassword,
            serverPort = serverPort,
            prefix = prefix,
            websocketEnabled = websocketEnabled,
            tcpPath = tcpPath,
            udpPath = udpPath
        )
        logger.log("Outline URL built (prefix=${prefix.isNotEmpty()}, ws=$websocketEnabled, tcpPath=${tcpPath.isNotEmpty()}, udpPath=${udpPath.isNotEmpty()})")
        logger.log("Outline URL: $outlineUrl")

        dobbyVpnService?.run {
            logger.log("[svc:$${serviceId}] setupVpn(): begin")
            vpnInterface = runCatching {
                interfaceFactory.create(context=this, vpnService=this).establish()
            }.onFailure { e ->
                logger.log("[svc:$${serviceId}] setupVpn(): establish FAILED: ${e.message}")
            }.getOrNull()

        }

        val tunFd = fdManager.GetTunFd(serviceId, dobbyVpnService)
        if (tunFd < 0) return false

        logger.log("[svc:$serviceId] startCloakOutline(): initializing Outline with tunFd=$tunFd")

        val connected = outlineLibFacade.init(outlineUrl, tunFd)
        if (!connected) {
            logger.log("Outline connection FAILED, stopping VPN service")
            return false
        }
        logger.log("outlineLibFacade connected successfully")
        if (websocketEnabled) {
            logger.log("WebSocket transport connected successfully")
        }
        logger.log("[svc:$serviceId] startCloakOutline(): completed (status=true) vpnInterface=${dobbyVpnService?.vpnInterface?.fd}")
        return true
    }

    fun stopOutline() {
        outlineLibFacade.disconnect()
    }
}
